src/lib/broadcast.jl

```diff
@@ -164,31 +164,46 @@ end
 # Avoid hitting special cases for `Adjoint` etc.
 _broadcast(f::F, x...) where F = materialize(broadcasted(f, x...))
 
-_get(x::Tuple, i) = x[i]
-_get(::Nothing, i) = nothing
 collapse_nothings(xs::AbstractArray{Nothing}) = nothing
 collapse_nothings(xs) = xs
 
-@adjoint function broadcasted(::AbstractArrayStyle, f, args...)
+_purefun(::Type{F}) where {F<:Function} = isempty(fieldnames(F))
+_purefun(::Type) = false
+if VERSION >= v"1.6"
+  _purefun(::Type{ComposedFunction{F,G}}) where {F,G} = _purefun(F) && _purefun(G)
+end
+_purefun(::Type{typeof(^)}) = false  # fix @testset "power" & @testset "diagonal hessian"
```

Comment:
This is a taken as a necessary condition for using ForwardDiff. The one suggested in #336 may also work fine: on examples from the tests:
```
julia> Base.issingletontype(typeof(identity∘tanh))
true

julia> let y = [1,2]
       Base.issingletontype(typeof(z->z^2+y[1]))
       end
false
```
Maybe there are other cases where these differ?